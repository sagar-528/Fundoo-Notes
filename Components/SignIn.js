import React from 'react'
import {  View, Text, TextInput, StyleSheet, TouchableOpacity, Keyboard, TouchableWithoutFeedback } from 'react-native';
import * as Animatable from 'react-native-animatable';
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5';
import Feather from 'react-native-vector-icons/Feather';
import LinearGradient from 'react-native-linear-gradient';

const SignIn = () => {
  return(
    <TouchableWithoutFeedback onPress={() => {
        Keyboard.dismiss();
        console.log('dismissed keyboard');
    }}>
    <View style={styles.container}>

        <View style={styles.header}>
            <Text style={styles.text_header}>Welcome!!!</Text>
        </View>

        <Animatable.View 
            animation="fadeInUpBig"
            style={styles.footer}
        >
            <Text style={styles.text_footer}>Username</Text>
            
            <View style={styles.action}>
                <FontAwesome5 
                    name="user"
                    size={20}
                    color="#fff"
                />
                <TextInput 
                    placeholder="Your Username"
                    placeholderTextColor="#fff"
                    style={styles.textInput}
                />
            </View>
            
            <Text style={[styles.text_footer, {
                marginTop: 35
            }]}>Password</Text>
            
            <View style={styles.action}>
                <Feather 
                    name="lock"
                    size={20}
                    color="#fff"
                />
                <TextInput 
                    placeholder="Your Password"
                    placeholderTextColor="#fff"
                    style={styles.textInput}
                 />
            </View>
                <TouchableOpacity>
                    <Text style={{color: 'black', marginTop:15, fontSize: 15}}>Forgot password?</Text>
                </TouchableOpacity>
            
                <View style={styles.button}>
                <TouchableOpacity
                    style={styles.signIn}>
                <LinearGradient
                    colors={['#08d4c4', '#01ab9d']}
                    style={styles.signIn}
                >
                <Text style={[styles.textSign, {
                    color:'#fff'
                }]}>Sign In</Text>
                </LinearGradient>
                </TouchableOpacity>

                <TouchableOpacity
                    style={[styles.signIn, {
                        borderColor: '#009387',
                        borderWidth: 1,
                        marginTop: 15
                    }]}
                >
                    <Text style={[styles.textSign, {
                        color: '#fff'
                    }]}>Sign Up</Text>
                </TouchableOpacity>
            </View>
        </Animatable.View>
    </View>
    </TouchableWithoutFeedback>
  );
};

export default SignIn

const styles = StyleSheet.create({
    container: {
        flex: 1, 
        backgroundColor: '#e0e0e0'
      },
      header: {
          flex: 1,
          justifyContent: 'flex-end',
          paddingHorizontal: 20,
          paddingBottom: 50
      },
      footer: {
          flex: 3,
          backgroundColor: '#616161',
          borderTopLeftRadius: 30,
          borderTopRightRadius: 30,
          paddingHorizontal: 20,
          paddingVertical: 30
      },
      text_header: {
        color: '#616161',
        fontWeight: 'bold',
        fontSize: 30
    },
    textInput: {
        flex: 1,
        marginTop: Platform.OS === 'ios' ? 0 : -12,
        paddingLeft: 10,
        color: '#05375a',
    },
    text_footer: {
        color: '#009387',
        fontSize: 25
    },
    action: {
        flexDirection: 'row',
        marginTop: 10,
        borderBottomWidth: 1,
        borderBottomColor: '#f2f2f2',
        paddingBottom: 5
    },
    button: {
        alignItems: 'center',
        marginTop: 45
    },
    signIn: {
        width: '100%',
        height: 50,
        justifyContent: 'center',
        alignItems: 'center',
        borderRadius: 10
    },
    textSign: {
        fontSize: 18,
        fontWeight: 'bold'
    }
})
