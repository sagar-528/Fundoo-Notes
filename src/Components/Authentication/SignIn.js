import React from 'react'
import {  View, Text, TextInput, TouchableOpacity, Keyboard, TouchableWithoutFeedback } from 'react-native';
import * as Animatable from 'react-native-animatable';
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5';
import Feather from 'react-native-vector-icons/Feather';
import GlobalCss from '../../Styles/GlobalCss';
import UserServices from '../../../Service/UserServices';
import UserSocialServices from '../../../Service/UserSocialServices';
import AsyncStorage from '@react-native-async-storage/async-storage';
import * as Keychain from 'react-native-keychain'
import { connect } from 'react-redux'
import SQLiteLabelServices from '../../../Service/SQLiteLabelServices'
import {storeUserID, storeUserLabel} from '../../Redux/Actions/CreateNewLabelActions'
import NoteDataControllerServices from '../../../Service/NoteDataControllerServices';
import SQLiteServices from '../../../Service/SQLiteServices'

class SignIn extends React.Component {

constructor(props) {
    super(props)

    this.state = {
        email : '',
        password : '',
        invalidEmail : false,
        invalidPassword : false,
        secureTextPassword : true,
        emailEmpty : false,
        passwordEmpty : false,
        isLoggedIn : false
    }
}

async componentDidMount(){
    try {
        const isLoggedIn = JSON.parse(await AsyncStorage.getItem('isLoggedIn'))
        if(isLoggedIn) {
            const credential = await Keychain.getGenericPassword();
            const UserCredential = JSON.parse(credential.password);
            await NoteDataControllerServices.retrieveDataFromFirebase(UserCredential.user.uid)

            this.props.storeUserId(UserCredential.user.uid)
            this.storeUserLabels(UserCredential.user.uid)
            this.props.navigation.push('Home', { screen: 'Notes' })
        }
      } 
      catch(e) {
        console.log(e)
      }
}

storeUserLabels = (userId) => {
    SQLiteLabelServices.selectLabelFromSQliteStorage(userId)
        .then(async result => {
            var temp = [];
            if(result.rows.length != 0) {
                for (let i = 0; i < result.rows.length; ++i)
                    temp.push(result.rows.item(i));
                } 
                this.props.storeUserLabel(temp)                 
        })
        .catch(error => console.log(error))
}

textInputChangeEmail = (val) =>{
     this.setState({
          email : val,
          invalidEmail : false,
          invalidPassword : false,
          emailEmpty : false
      })
  }
  
  textInputChangePassword = (val) =>{
     this.setState({
          password : val,
          invalidEmail : false,
          invalidPassword : false,
          emailEmpty : false
      })
  }

handleSecureTextPassword = () => {
    const {onPress} = this.props
    if(this.state.secureTextPassword == true) {
        this.setState({
            secureTextPassword : false
        })
    }
    else {
        this.setState({
            secureTextPassword : true
        })
    }
    (onPress == undefined) ? null : onPress();
}

signInHandler = async () => {
    const {onPress} = this.props
    if(this.state.email != '' && this.state.password != '')
    {
        await UserServices.SignIn(this.state.email, this.state.password)
                .then( async (UserCredential) => {   
                        this.storeIteminAsyncStorage()
                        console.log("signIn"); 
                        await NoteDataControllerServices.retrieveDataFromFirebase(UserCredential.user.uid) 
                        await Keychain.setGenericPassword('UserCredential', JSON.stringify(UserCredential));
                        
                        await this.props.storeUserId(UserCredential.user.uid)
                        await this.storeUserLabels(UserCredential.user.uid)
                        this.props.navigation.navigate('Home', { screen: 'Notes' })
                    })
                .catch(error => {
                    if(error == 'Email not Found') {
                        this.setState({
                        invalidEmail : true
                        })
                    }
                    if(error == 'Incorrect Password') {
                        this.setState({
                            invalidPassword : true
                        })
                        console.log('incorrect password');
                    }
                })
        }
    else{
        if(this.state.email == '') {
           this.setState({
                emailEmpty : true
            })
        }
        if(this.state.password == '') {
             this.setState({
                passwordEmpty : true
            })
        }
    }
    // onPress();
}

storeIteminAsyncStorage = () => {
    try {
         this.setState({
            isLoggedIn : true
        })
         AsyncStorage.setItem('isLoggedIn', JSON.stringify(this.state.isLoggedIn));
    } catch (e) {
        console.log(e);
    }
}

SignUpHandler = () =>{
    const {onPress} = this.props;
    this.props.navigation.navigate('SignUp')
    // onPress();
}

forgotPasswordHandler = () =>{
    const {onPress} = this.props;
    this.props.navigation.navigate('ForgotPassword')
    // onPress();
}

handleFacebookLoginButton = async () => {
    const {onPress} = this.props;
    UserSocialServices.facebookLogin()
    .then( async UserCredential => {

        UserServices.readUserDataFromRealtimeDatabase(UserCredential.user.uid)
            .then(data =>{
                if(data == null){
                    UserServices.writeUserDataInRealtimeDatabase(
                        UserCredential.user.uid, 
                        UserCredential.additionalUserInfo.profile.first_name, 
                        UserCredential.additionalUserInfo.profile.last_name, 
                        UserCredential.additionalUserInfo.profile.email);
                }
            })
       
        await NoteDataControllerServices.retrieveDataFromFirebase(UserCredential.user.uid)
        this.storeIteminAsyncStorage()
        await Keychain.setGenericPassword('UserCredential', JSON.stringify(UserCredential));
        
        await this.props.storeUserId(UserCredential.user.uid)
        await this.storeUserLabels(UserCredential.user.uid)
        this.props.navigation.navigate('Home', { screen: 'Notes' })
    })
        .catch(error => {
            console.log(error)
        })
    // onPress();
}

render(){
  return(
    <TouchableWithoutFeedback onPress = {() => {
        Keyboard.dismiss();
    }}>
    <View style = {GlobalCss.container}>
        <View style = {GlobalCss.header}>
            <FontAwesome5 
                name = "foursquare"
                color = "#fff"
                size = {50}
            />
            <Text style = {GlobalCss.text_header}>Fundoo Notes</Text>
        </View>

        <Animatable.View 
            animation = "fadeInUpBig"
            style = {GlobalCss.footer}
        >
            <Text style = {GlobalCss.text_footer}>Email-Id</Text>
            
            <View style = {GlobalCss.action}>
                <FontAwesome5 
                    name = "user"
                    size = {20}
                    color = "black"
                />
                <TextInput 
                    placeholder = "Your Email-Id"
                    placeholderTextColor = "black"
                    style = {GlobalCss.textInput}
                    onChangeText={this.textInputChangeEmail}
                />
            </View>
            <View>
                <Text style={GlobalCss.text_error_style}>
                    {(this.state.emailEmpty) ? 'Required..' : (this.state.invalidEmail) ? 'Email not Found..' : null}
                </Text>
            </View>
            
            <Text style = {[GlobalCss.text_footer, {
                marginTop: 10
            }]}>Password</Text>
            
            <View style = {GlobalCss.action}>
                <Feather 
                    name = "lock"
                    size = {20}
                    color = "black"
                />
                <TextInput 
                    placeholder = "Your Password"
                    placeholderTextColor = "black"
                    style = {GlobalCss.textInput}
                    onChangeText={this.textInputChangePassword}
                    secureTextEntry = {this.state.secureTextPassword}
                 />
            </View>
            <View>
                <Text style = {GlobalCss.text_error_style}>
                    {(this.state.passwordEmpty) ? 'Required..' : (this.state.invalidPassword) ? 'Invalid Password..' : null}
                </Text>
            </View>
                <TouchableOpacity onPress = { this.forgotPasswordHandler }>
                    <Text style = {{color: '#1976d2', marginTop:5, fontSize: 15}}>Forgot password?</Text>
                </TouchableOpacity>
            
                <View style = {GlobalCss.button}>
                <TouchableOpacity
                    style = {[GlobalCss.signIn, {
                        borderColor: '#009387',
                        borderWidth: 1,
                        marginTop: 15
                    }]}
                    onPress = { this.signInHandler }
                >
                <Text style = {[GlobalCss.textSign, {
                    color:'black'
                }]}>Sign In</Text>
                </TouchableOpacity>

                <TouchableOpacity
                    style = {[GlobalCss.signIn, {
                        borderColor: '#009387',
                        borderWidth: 1,
                        marginTop: 15
                    }]}
                    onPress = { this.SignUpHandler }
                >
                    <Text style = {[GlobalCss.textSign, {
                        color: 'black'
                    }]}>Sign Up</Text>
                </TouchableOpacity>
            </View>
                <View>
                    <Text style={{marginTop: 10, textAlign: 'center', color: 'black', fontWeight: 'bold', fontSize: 15,}}>Or</Text>
                </View>
                <TouchableOpacity onPress={this.handleFacebookLoginButton}>
                <View style={GlobalCss.facebookButton} >
                   <FontAwesome5 
                   name = "facebook"
                   size = {20}
                   color = "blue" />
                   <Text style = {[GlobalCss.homeTextSign,{
                    color: '#1976d2'
                   }]}>Login with Facebook</Text>
            </View>
            </TouchableOpacity>
    </Animatable.View>
    </View>
    </TouchableWithoutFeedback>
    )};
};

const mapStateToProps = state => {
    return {
        userId : state.createLabelReducer.userId,
        userLabel : state.createLabelReducer.userLabel
    }
  }

  const mapDispatchToProps = dispatch => {
    return {
        storeUserId : (userId) => dispatch(storeUserID(userId)),
        storeUserLabel : (userLabel) => dispatch(storeUserLabel(userLabel))
    }
  }

export default connect(mapStateToProps,mapDispatchToProps)(SignIn)
