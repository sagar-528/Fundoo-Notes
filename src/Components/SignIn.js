import React from 'react'
import {  View, Text, TextInput, TouchableOpacity, Keyboard, TouchableWithoutFeedback } from 'react-native';
import * as Animatable from 'react-native-animatable';
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5';
import Feather from 'react-native-vector-icons/Feather';
import GlobalCss from '../Styles/GlobalCss';
import UserServices from '../../Service/UserServices';
import UserSocialServices from '../../Service/UserSocialServices';


class SignIn extends React.Component {

constructor(props) {
    super(props)

    this.state = {
        email : '',
        password : '',
        invalidEmail : false,
        invalidPassword : false,
        secureTextPassword : true,
        emailEmpty : false,
        passwordEmpty : false,
    }
}


textInputChangeEmail = (val) =>{
     this.setState({
          email : val,
          invalidEmail : false,
          invalidPassword : false,
          emailEmpty : false
      })
  }
  
  textInputChangePassword = (val) =>{
     this.setState({
          password : val,
          invalidEmail : false,
          invalidPassword : false,
          emailEmpty : false
      })
  }

handleSecureTextPassword = () => {
    const {onPress} = this.props
    if(this.state.secureTextPassword == true) {
        this.setState({
            secureTextPassword : false
        })
    }
    else {
        this.setState({
            secureTextPassword : true
        })
    }
    (onPress == undefined) ? null : onPress();
}

signInHandler = () => {
    const {onPress} = this.props
    if(this.state.email != '' && this.state.password != '')
    {
                UserServices.SignIn(this.state.email, this.state.password)
                .then(UserCredential => {
                        console.log("signIn");
                        this.props.navigation.push('Home')
                    })
                .catch(error => {
                    if(error == 'Email not Found') {
                        this.setState({
                        invalidEmail : true
                        })
                    }
                    if(error == 'Incorrect Password') {
                        this.setState({
                            invalidPassword : true
                        })
                        console.log('incorrect password');
                    }
                })
        }
    else{
        if(this.state.email == '') {
           this.setState({
                emailEmpty : true
            })
        }
        if(this.state.password == '') {
             this.setState({
                passwordEmpty : true
            })
        }
    }
    onPress();
}

SignUpHandler = () =>{
    const {onPress} = this.props;
    this.props.navigation.navigate('SignUp')
    onPress();
}

forgotPasswordHandler = () =>{
    const {onPress} = this.props;
    this.props.navigation.navigate('ForgotPassword')
    onPress();
}

handleFacebookLoginButton = () => {
    UserSocialServices.facebookLogin()
    .then(UserCredential => {
        UserSocialServices.writeUserDataForFacebookLogin(UserCredential);
        this.props.navigation.push('Home')
    })
        .catch(error => {
            console.log(error)
        })
}

render(){
  return(
    <TouchableWithoutFeedback onPress = {() => {
        Keyboard.dismiss();
    }}>
    <View style = {GlobalCss.container}>
        <View style = {GlobalCss.header}>
            <FontAwesome5 
                name = "foursquare"
                color = "#fff"
                size = {50}
            />
            <Text style = {GlobalCss.text_header}>Fundoo Notes</Text>
        </View>

        <Animatable.View 
            animation = "fadeInUpBig"
            style = {GlobalCss.footer}
        >
            <Text style = {GlobalCss.text_footer}>Email-Id</Text>
            
            <View style = {GlobalCss.action}>
                <FontAwesome5 
                    name = "user"
                    size = {20}
                    color = "black"
                />
                <TextInput 
                    placeholder = "Your Email-Id"
                    placeholderTextColor = "black"
                    style = {GlobalCss.textInput}
                    onChangeText={this.textInputChangeEmail}
                />
            </View>
            <View>
                <Text style={GlobalCss.text_error_style}>
                    {(this.state.emailEmpty) ? 'Required..' : (this.state.invalidEmail) ? 'Email not Found..' : null}
                </Text>
            </View>
            
            <Text style = {[GlobalCss.text_footer, {
                marginTop: 10
            }]}>Password</Text>
            
            <View style = {GlobalCss.action}>
                <Feather 
                    name = "lock"
                    size = {20}
                    color = "black"
                />
                <TextInput 
                    placeholder = "Your Password"
                    placeholderTextColor = "black"
                    style = {GlobalCss.textInput}
                    onChangeText={this.textInputChangePassword}
                    secureTextEntry = {this.state.secureTextPassword}
                 />
            </View>
            <View>
                <Text style = {GlobalCss.text_error_style}>
                    {(this.state.passwordEmpty) ? 'Required..' : (this.state.invalidPassword) ? 'Invalid Password..' : null}
                </Text>
            </View>
                <TouchableOpacity onPress = { this.forgotPasswordHandler }>
                    <Text style = {{color: '#1976d2', marginTop:5, fontSize: 15}}>Forgot password?</Text>
                </TouchableOpacity>
            
                <View style = {GlobalCss.button}>
                <TouchableOpacity
                    style = {[GlobalCss.signIn, {
                        borderColor: '#009387',
                        borderWidth: 1,
                        marginTop: 15
                    }]}
                    onPress = { this.signInHandler }
                >
                <Text style = {[GlobalCss.textSign, {
                    color:'black'
                }]}>Sign In</Text>
                </TouchableOpacity>

                <TouchableOpacity
                    style = {[GlobalCss.signIn, {
                        borderColor: '#009387',
                        borderWidth: 1,
                        marginTop: 15
                    }]}
                    onPress = { this.SignUpHandler }
                >
                    <Text style = {[GlobalCss.textSign, {
                        color: 'black'
                    }]}>Sign Up</Text>
                </TouchableOpacity>
            </View>
                <View>
                    <Text style={{marginTop: 10, textAlign: 'center', color: 'black', fontWeight: 'bold', fontSize: 15,}}>Or</Text>
                </View>
                <TouchableOpacity onPress={this.handleFacebookLoginButton}>
                <View style={GlobalCss.facebookButton} >
                   <FontAwesome5 
                   name = "facebook"
                   size = {20}
                   color = "blue" />
                   <Text style = {[GlobalCss.homeTextSign,{
                    color: '#1976d2'
                   }]}>Login with Facebook</Text>
            </View>
            </TouchableOpacity>
    </Animatable.View>
    </View>
    </TouchableWithoutFeedback>
  )};
};

export default SignIn
