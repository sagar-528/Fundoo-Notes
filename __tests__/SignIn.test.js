import React from 'react';
import {shallow} from 'enzyme';
import SignIn from '../src/Components/SignIn'

describe('test SignIn', () => {
    it('test when render should match to snapshot', () => {
        const component = shallow(<SignIn/>)
        expect(component).toMatchSnapshot();
    })

    it('test the email handler method should update email state', () => {
        const component = shallow(<SignIn/>)
        expect(component.instance().state.email).toBe('')
        component.instance().textInputChangeEmail('gupta.sagar528@gmail.com')
        expect(component.instance().state.email).toBe('gupta.sagar528@gmail.com')
    })

    it('test the password handler method should update password state', async () => {
        const component = shallow(<SignIn/>)
        expect(component.instance().state.password).toBe('')
        component.instance().textInputChangePassword('Qwerty@111')
        expect(component.instance().state.password).toBe('Qwerty@111')
    })

    it('test onPress event of eye icon of password textinput called it will change the secureTextPassword State', () => {
        const onPressEvent = jest.fn();
        const component = shallow(<SignIn onPress = {onPressEvent}/>)
        const instance = component.instance();
        expect(instance.state.secureTextPassword).toBe(true);
    })

    it('test onPress event of sign up button it will navigate to sign up screen', () => {
        const navigation = { navigate : jest.fn() }
        const onPressEvent = jest.fn();
        const component = shallow(<SignIn onPress = {onPressEvent} navigation = {navigation}/>)
        const instance = component.instance();
        instance.SignUpHandler();

        expect(onPressEvent).toHaveBeenCalled();
        expect(navigation.navigate).toBeCalledWith('SignUp');
    })

    it('test onPress event of forgot password button it will navigate forgot password screen', () => {
        const navigation = { navigate : jest.fn() }
        const onPressEvent = jest.fn();
        const component = shallow(<SignIn onPress = {onPressEvent} navigation = {navigation}/>)
        const instance = component.instance();
        instance.forgotPasswordHandler();
        expect(onPressEvent).toHaveBeenCalled();
        expect(navigation.navigate).toBeCalledWith('ForgotPassword');
    })
})