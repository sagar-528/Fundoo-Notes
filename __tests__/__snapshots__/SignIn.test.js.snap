// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Login test when render should match to snapshot 1`] = `ShallowWrapper {}`;

exports[`test SignIn test when render should match to snapshot 1`] = `ShallowWrapper {}`;
